### PR TITLE
Library categories list api created

### PR DESCRIPTION
**Pull request by aman-tunerlabs**

### CODE REVIEW

File: migrations/20200526095359-create-program-for-observations.js [hunk 1]
```diff
@@ -1,98 +1,203 @@
...
+
 module.exports = {
   async up(db) {
     
     global.migrationMsg = "Programs added for observations and observation submissions";
     
+    // Dependencies
+    const csv = require("csvtojson");
+    const fs = require("fs");
+    let path = "public/csv/program-in-observations.csv";
```
Comment (aks30) [inline migrations/20200526095359-create-program-for-observations.js:11]: @aman-tunerlabs - Put the path as project root, and file name as observationSolutionToProgramMapping

File: migrations/20200623141837-upload-library-image-in-gcp.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
+module.exports = {
+  async up(db) {
+    
+    const { Storage } = require("@google-cloud/storage");
```
Comment (aks30) [inline migrations/20200623141837-upload-library-image-in-gcp.js:4]: @aman-tunerlabs - Make a separate file for image upload to cloud, also it should work with azure or aws or google cloud

File: migrations/20200623141837-upload-library-image-in-gcp.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    const fs = require("fs");
+    global.migrationMsg = "Upload library images";
+
+    let libraryImageFolder = "public/images/library";
+
+    const storage = new Storage({
+      projectId : "shikshalokam",
+      keyFilename : "generics/helpers/credentials/sl-dev-storage.json"
```
Comment (aks30) [inline migrations/20200623141837-upload-library-image-in-gcp.js:12]: @aman-tunerlabs - The path has to be picked from env

File: migrations/20200623141837-upload-library-image-in-gcp.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    let images = [];
+
+    fs.readdirSync(libraryImageFolder).forEach(function (file) {
+      images.push(file)
+    });
+
+    let libraryCategories = [];
+
+    for( let image = 0; image < images.length ; image++ ) {
```
Comment (aks30) [inline migrations/20200623141837-upload-library-image-in-gcp.js:25]: @aman-tunerlabs - Create static objects and inert as required in DB

File: models/libraryCategories.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+/**
+ * name : libraryCategories.js
+ * author : Aman
+ * created-date : 23-Jun-2020
+ * Description : Library categories.
+ */
+
+module.exports = {
+    name : "libraryCategories",
+    schema: {
```
Comment (aks30) [inline models/libraryCategories.js:10]: @aman-tunerlabs - Add proper validations and default values

File: migrations/20200623141837-upload-library-image-in-gcp.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        "updatedAt" : new Date(),
+        "createdBy" : "SYSTEM",
+        "updatedBy" : "SYSTEM",
+        "isDeleted" : false,
+        "isVisible" : true,
+        "status" : "active"        
+      });
+    }
+
+    await db.collection('libraryCategories').insertMany(libraryCategories);
```
Comment (aks30) [inline migrations/20200623141837-upload-library-image-in-gcp.js:54]: @aman-tunerlabs - Add index for the new collection on externalId as unique

File: module/assessments/helper.js [hunk 7]
```diff
@@ -527,10 +528,58 @@ module.exports = class AssessmentsHelper {
...
+              await database.models.entityAssessors.create({
+                  userId : userDetails.userId,
+                  email : userDetails.email,
+                  name : userDetails.name,
+                  externalId : userDetails.externalId,
+                  programId : createdSolutionAndProgram.programId,
+                  solutionId : createdSolutionAndProgram._id,
+                  entityTypeId : createdSolutionAndProgram.entityTypeId,
+                  entityType : createdSolutionAndProgram.entityType,
+                  role : messageConstants.common.ASSESSOR,
```
Comment (aks30) [inline module/assessments/helper.js:545]: @aman-tunerlabs - Make the role as LEAD_ASSESSOR

File: module/assessments/helper.js [hunk 8]
```diff
@@ -527,10 +528,58 @@ module.exports = class AssessmentsHelper {
...
+                        entityName : entity.metaInformation.name,
+                        programId : createdSolutionAndProgram.programId,
+                        solutionName : createdSolutionAndProgram.name,
+                        solutionId : createdSolutionAndProgram._id,
+                        programName : createdSolutionAndProgram.programName,
+                        solutionSubType : createdSolutionAndProgram.subType
+                    }
+                });
+
+                await entityAssessorsHelper.sendUserNotifications(
```
Comment (aks30) [inline module/assessments/helper.js:577]: @aman-tunerlabs - No need to send user a notification in this case

File: module/assessments/helper.js [hunk 9]
```diff
@@ -527,10 +528,58 @@ module.exports = class AssessmentsHelper {
...
+                  solutionId : createdSolutionAndProgram._id,
+                  entityTypeId : createdSolutionAndProgram.entityTypeId,
+                  entityType : createdSolutionAndProgram.entityType,
+                  role : messageConstants.common.ASSESSOR,
+                  createdBy : userDetails.userId,
+                  updatedBy : userDetails.userId,
+                  entities : createdSolutionAndProgram.entities
+              });
+
+              if(
```
Comment (aks30) [inline module/assessments/helper.js:551]: @aman-tunerlabs - Not required this block of code

File: module/solutions/helper.js [hunk 10]
```diff
@@ -913,7 +914,12 @@ module.exports = class SolutionsHelper {
                     "programId",
                     "entityTypeId",
                     "entityType",
-                    "isAPrivateProgram"
+                    "isAPrivateProgram",
+                    "entities",
```
Comment (aks30) [inline module/solutions/helper.js:918]: @aman-tunerlabs - No longer required as we don't need to send notifications

File: module/library/categories/helper.js [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+                    
+                    categories[category.icon] = {
+                        name : category.name,
+                        type : category.externalId
+                    };
+
+                    return category.icon
+                });
+
+                let downloadableImages = await kendraService.gcpDownloadableImage(
```
Comment (aks30) [inline module/library/categories/helper.js:113]: @aman-tunerlabs - It has to work with azure or aws or gcp

File: module/library/categories/helper.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+                );
+
+                if( downloadableImages.status !== httpStatusCode.ok.status) {
+                    throw {
+                        status : httpStatusCode.bad_request.status,
+                        message : messageConstants.apiResponses.URL_COULD_NOT_BE_FOUND
+                    }
+                }
+
+                downloadableImages = downloadableImages.result.map(downloadableImage=>{
```
Comment (aks30) [inline module/library/categories/helper.js:127]: @aman-tunerlabs - The categroy list API response can be cached for faster retrieval as it not going to change often

File: generics/services/kendra.js [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+/**
+  * Get downloadable file.
+  * @function
+  * @name gcpDownloadFile
+  * @param {Object} bodyData - body data.
+  * @returns {Array} GCP Downloadable file.
+*/
+
+var gcpDownloadFile = function (bodyData) {
```
Comment (aks30) [inline generics/services/kendra.js:20]: @aman-tunerlabs - Function can be a const and not var. Also think of a common function here, just the endpoint switches based on selected cloud storage in env. Can be a single function with switch case on cloud storage

File: config/globalVariable.js [hunk 14]
```diff
@@ -111,6 +110,86 @@ module.exports = function () {
...
+
+        return category.icon
+    });
+
+    let kendraService = require(ROOT_PATH + "/generics/services/kendra");
+
+    let downloadableImages = new Promise(async function(resolve, reject) {
+
+      let image = "";
+      if( process.env.CLOUD_STORAGE === "GC" ) {
```
Comment (aks30) [inline config/globalVariable.js:148]: @aman-tunerlabs - Make it a single function in Kendra service, let the cloud storage switch case be there.

File: config/globalVariable.js [hunk 15]
```diff
@@ -111,6 +110,86 @@ module.exports = function () {
     }
   });
 
+  global.sessions = {};
+
+  let categories = new Promise(async function(resolve, reject) {
```
Comment (aks30) [inline config/globalVariable.js:115]: @aman-tunerlabs - Let's use library categories helper and not write DB queries here.

File: imageUploadToCloud.js [hunk 16]
```diff
@@ -0,0 +1,94 @@
+const { Storage } = require("@google-cloud/storage");
```
Comment (aks30) [inline imageUploadToCloud.js:1]: @aman-tunerlabs - Why not use Kendra service for the uploads, why create a new file ?

File: migrations/20200526095359-create-program-for-observations.js [hunk 17]
```diff
@@ -1,98 +1,203 @@
...
+
 module.exports = {
   async up(db) {
     
     global.migrationMsg = "Programs added for observations and observation submissions";
     
+    // Dependencies
+    const csv = require("csvtojson");
+    const fs = require("fs");
+    let path = "/observationSolutionToProgramMapping.csv";
```
Comment (aks30) [inline migrations/20200526095359-create-program-for-observations.js:11]: @aman-tunerlabs - This should be in project root not in system or machine root path

File: migrations/20200623141837-upload-library-image-in-gcp.js [hunk 18]
```diff
@@ -0,0 +1,76 @@
+module.exports = {
+  async up(db) {
+    
+    const uploadImage = require("../imageUploadToCloud.js");
```
Comment (aks30) [inline migrations/20200623141837-upload-library-image-in-gcp.js:4]: @aman-tunerlabs - Let's use kendra to upload images please.

File: module/assessments/helper.js [hunk 19]
```diff
@@ -508,15 +509,60 @@ module.exports = class AssessmentsHelper {
...
+                      "entityType",
+                      "entityTypeId"
+                    ]
+              )
+
+              if( solutionData[0].subType === messageConstants.common.INDIVIDUAL ) {
+
+                    let entity = 
+                    await entitiesHelper.entityDocuments({
+                        "metaInformation.externalId" :  userDetails.userId
```
Comment (aks30) [inline module/assessments/helper.js:536]: @aman-tunerlabs - "metaInformation.externalId" is not indexed so it should not be a query key, would suggest having a user extension reference and keycloak userId in the document root in these cases and use that. 
Filter here should be user id in combination with the entity type.

File: module/assessments/helper.js [hunk 20]
```diff
@@ -508,15 +509,60 @@ module.exports = class AssessmentsHelper {
...
+                    },["_id"])
+
+                    let entityId;
+
+                    if( !entity[0] ) {
+                        
+                        let entity = await database.models.entities.create({
```
Comment (aks30) [inline module/assessments/helper.js:543]: @aman-tunerlabs - Use helper function here please

File: module/assessments/helper.js [hunk 21]
```diff
@@ -527,10 +573,25 @@ module.exports = class AssessmentsHelper {
               await solutionsHelper.createProgramAndSolutionFromTemplate(
                 templateId,
                 requestedData.program,
-                userId,
+                userDetails.userId,
                 solutionInformation,
                 true
               );
+
+              await database.models.entityAssessors.create({
```
Comment (aks30) [inline module/assessments/helper.js:581]: @aman-tunerlabs - Use helper functions please

File: module/library/categories/helper.js [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+                        
+                        categories[category.icon] = {
+                            name : category.name,
+                            type : category.externalId
+                        };
+    
+                        return category.icon
+                    });
+    
+                    if( process.env.CLOUD_STORAGE === "GC" ) {
```
Comment (aks30) [inline module/library/categories/helper.js:127]: @aman-tunerlabs - Call one kendra function , let the switch case be there.

File: module/library/categories/helper.js [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+
+                } else {
+                    
+                    let libraryCategories = 
+                    await this.categoryDocuments(
+                        {
+                            status : messageConstants.common.ACTIVE_STATUS
+                        },
+                        [
+                            "externalId",
```
Comment (aks30) [inline module/library/categories/helper.js:102]: @aman-tunerlabs - Give updated At in response, app can use it decide or not to refresh their local storage.

File: imageUploadToCloud.js (hunk 16 above)
Comment (aman-tunerlabs) [reply imageUploadToCloud.js:1]: @aks30 Earlier Comment says to create a seperate file for image upload to cloud. But latest comment says to use kendra service for upload.

File: module/assessments/helper.js (hunk 19 above)
Comment (aman-tunerlabs) [reply module/assessments/helper.js:536]: @aks30 Did not got this man .

File: generics/services/kendra.js [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+const upload = function (file,filePath,storage) {
+
+    let fileUploadUrl = process.env.KENDRA_APPLICATION_ENDPOINT; 
+    let bucketName = "";
+
+    if ( storage === "GC" ) {
+        fileUploadUrl = fileUploadUrl + "api/v1/cloud-services/gcp/uploadFile";
+        bucketName = process.env.GCP_BUCKET_NAME;
+    } else if( storage === "AWS" ) {
+        console.log("here")
```
Comment (aks30) [inline generics/services/kendra.js:82]: @aman-tunerlabs - Remove console logs.

File: migrations/20200623141837-upload-library-image-in-gcp.js [hunk 25]
```diff
@@ -0,0 +1,78 @@
+module.exports = {
+  async up(db) {
+    
+    const kendraServiceHelper = require("../generics/services/kendra");
```
Comment (aks30) [inline migrations/20200623141837-upload-library-image-in-gcp.js:4]: @aman-tunerlabs - Rename file, remove gcp

File: generics/services/kendra.js [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+  * @param {Object} bodyData - body data.
+  * @returns {Array} upload file.
+*/
+
+const upload = function (file,filePath,storage) {
+
+    let fileUploadUrl = process.env.KENDRA_APPLICATION_ENDPOINT; 
+    let bucketName = "";
+
+    if ( storage === "GC" ) {
```
Comment (aks30) [inline generics/services/kendra.js:78]: @aman-tunerlabs - Implement same logic of cloud storage as above

File: migrations/20200623141837-upload-library-image-in-gcp.js [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+        status : "active"
+      }
+    ];
+
+    for( let category = 0 ; category < categories.length; category++ ) {
+
+      await kendraServiceHelper.upload(
+        `public/assets/library/${categories[category].icon}`,
+        `static/${categories[category].icon}`,
+        process.env.CLOUD_STORAGE
```
Comment (aks30) [inline migrations/20200623141837-upload-library-image-in-gcp.js:64]: @aman-tunerlabs - This is not required.

File: module/entities/helper.js [hunk 28]
```diff
@@ -1201,6 +1201,48 @@ module.exports = class EntitiesHelper {
...
+   * Create individual entity.
+   * @method 
+   * @name createIndividualEntity
+   * @param {String} userDetails - User details.
+   * @param {String} entityType - Type of entity.
+   * @param {String} entityTypeId - Entity type id.
+   * @returns {String} Create individual entity.
+   */
+
+  static createIndividualEntity(
```
Comment (aks30) [inline module/entities/helper.js:1214]: @aman-tunerlabs - Use the common add function in this file to create an entity

File: module/entityAssessors/helper.js [hunk 29]
```diff
@@ -831,4 +836,45 @@ module.exports = class EntityAssessorHelper {
...
+     * @param {String} programId - program id.
+     * @param {String} solutionId - solution id.
+     * @param {String} entityType - entity type.
+     * @param {String} entityTypeId - entity type id.
+     * @param {Array}  entities - entities 
+     * @param {Object} userDetails - logged in user details. 
+     * @returns {Object} created Entity assessor data.
+     */
+
+    static create(programId,solutionId,entityType,entityTypeId, entities, userDetails) {
```
Comment (aks30) [inline module/entityAssessors/helper.js:852]: @aman-tunerlabs - Use a common query to create an entity assessor as already done in this file.

File: module/library/search/helper.js [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+      * @returns {Object} Search library solutions.
+     */
+
+    static search( searchText,pageSize,pageNo ) {
+        return new Promise(async (resolve, reject) => {
+            try {
+
+                let matchQuery = {
+                    $match : {
+                        isReusable : true,
```
Comment (aks30) [inline module/library/search/helper.js:34]: @aman-tunerlabs - Make this and other fields as discussed in index

File: module/solutions/helper.js [hunk 31]
```diff
@@ -1039,5 +1043,51 @@ module.exports = class SolutionsHelper {
...
+     /**
+     * Add roles in solution.
+     * @method
+     * @name addRoles
+     * @param {String} solutionId - solution id.
+     * @param {Array} allRoles - roles assigned to solution.
+     * @returns {Object} Add roles in solution.
+     */
+
+    static addRoles(
```
Comment (aks30) [inline module/solutions/helper.js:1056]: @aman-tunerlabs -Rename it to addDefaultACL

File: generics/services/kendra.js [hunk 32]
```diff
@@ -0,0 +1,125 @@
...
+
+        } catch (error) {
+            return reject(error);
+        }
+    })
+
+}
+
+/**
+  * Upload urls.
```
Comment (aks30) [inline generics/services/kendra.js:66]: @aman-tunerlabs - Function name doesn't explain the use of the function

File: generics/services/kendra.js [hunk 33]
```diff
@@ -0,0 +1,125 @@
...
+  * @function
+  * @name upload
+  * @param {Object} bodyData - body data.
+  * @returns {Array} upload file.
+*/
+
+const upload = function (file,filePath) {
+
+    let fileUploadUrl = process.env.KENDRA_APPLICATION_ENDPOINT;
```
Comment (aks30) [inline generics/services/kendra.js:75]: @aman-tunerlabs - Line 75-87 ca be one common function for upload and download function

**Pull request merged**